source/common/http/filter/gzip_filter.cc

```diff
@@ -0,0 +1,282 @@
+    compressor_.flush(compressed_data_);
+  }
+
+  if (compressed_data_.length()) {
+    data.drain(n_data);
+    data.move(compressed_data_);
+    return Http::FilterDataStatus::Continue;
+  }
+
+  return Http::FilterDataStatus::StopIterationNoBuffer;
+}
+
+bool GzipFilter::isAcceptEncodingAllowed(HeaderMap& headers) const {
+  const Http::HeaderEntry* accept_encoding = headers.AcceptEncoding();
+  // TODO(gsagula): Since gzip is the only available content-encoding in Envoy at the moment,
+  // order/priority of preferred server encodings is disregarded (RFC2616-14.3). Replace this
+  // with a data structure that parses Accept-Encoding values and allows fast lookup of
+  // key/priority. Also, this should be part of some utility library.
+  if (accept_encoding) {
+    bool is_identity{true};  // true if does not exist or is followed by `q=0`.
+    bool is_wildcard{false}; // true if found and not followed by `q=0`.
+    bool is_gzip{false};     // true if exists and not followed by `q=0`.
+
+    std::vector<absl::string_view> tokens =
+        StringViewUtil::splitToken(headers.AcceptEncoding()->value().c_str(), ";,", false);
+
+    for (auto& token : tokens) {
+      absl::string_view next_token{};
+      absl::string_view trimmed_token = StringViewUtil::trim(token);
+      if (token != tokens.back()) {
+        next_token = StringViewUtil::trim(*(&token + 1));
+      }
+      if (trimmed_token == Http::Headers::get().AcceptEncodingValues.Gzip) {
+        is_gzip = next_token.empty() ? true : (next_token != ZeroQvalueString);
+      }
+      if (trimmed_token == Http::Headers::get().AcceptEncodingValues.Identity) {
+        is_identity = next_token.empty() ? false : (next_token == ZeroQvalueString);
+      }
+      if (trimmed_token == Http::Headers::get().AcceptEncodingValues.Wildcard) {
+        is_wildcard = !next_token.empty() ? true : (next_token != ZeroQvalueString);
+      }
+    }
+
+    return ((is_gzip && is_identity) || is_wildcard);
+  }
+
+  return true;
+}
+
+bool GzipFilter::isCacheControlAllowed(HeaderMap& headers) const {
+  const Http::HeaderEntry* cache_control = headers.CacheControl();
+  if (cache_control) {
+    return !StringViewUtil::findToken(cache_control->value().c_str(), ";=",
+                                      Http::Headers::get().CacheControlValues.NoTransform.c_str());
+  }
+  return true;
+}
+
+bool GzipFilter::isContentTypeAllowed(HeaderMap& headers) const {
+  const Http::HeaderEntry* content_type = headers.ContentType();
+  if (content_type && !config_->contentTypeValues().empty()) {
+    // TODO(gsagula): do this lookup without allocating a new string.
+    std::string value{StringViewUtil::cropRight(content_type->value().c_str(), ";")};
+    return config_->contentTypeValues().find(value) != config_->contentTypeValues().end();
+  }
+  return true;
+}
+
+bool GzipFilter::isEtagAllowed(HeaderMap& headers) const {
+  const Http::HeaderEntry* etag = headers.Etag();
+  if (config_->disableOnEtag()) {
```

Comment:
It's probably worth putting some comments here referencing all the controversy you found about how to handle etags with gzip.

Reply:
Sorry, I'm really not sure about Etags and gzip :(
I will add the comment.